File: Java/boj/Beginner/C_BinarySearch/CuttingRiceCake.java
package boj.Beginner.C_BinarySearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class CuttingRiceCake {
    private static StringTokenizer st;
    private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static List<Integer> arr;

    private static int cutter(int m) {
        int left = 0;
        int right = Collections.max(arr);
        int result = 0;

        while (left <= right) {
            int mid = (left + right) / 2;
            int tmp = 0;

            for (Integer t : arr)
                if (t > mid)
                    tmp += t - mid;

            if (tmp < m) right = mid - 1;
            else {
                result = mid;
                left = mid + 1;
            }
        }

        return result;
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        arr = new ArrayList<>();
        st = new StringTokenizer(br.readLine());

        for (int i = 0; i < n; i++)
            arr.add(Integer.parseInt(st.nextToken()));

        int result = cutter(m);
        System.out.println(result);
    }
}

File: v1/Python/beginner/03_BinarySearch/cuttingRiceCake.py
import sys

# 손님이 요청한 길이가 M일 때, 적어도 M만큼의 떡을 얻기 위해서
# 절단기에 설정할 수 있는 높이의 최댓값

input = sys.stdin.readline

def cutter(arr, m):
    left = 0
    right = max(arr)
    result = 0

    while left <= right:
        mid = (left + right) // 2
        tmp = 0

        for t in arr:
            if t > mid:
                tmp += t - mid
        
        if tmp < m:
            right = mid - 1
        else:
            left = mid + 1
            result = mid

    return result

if __name__ == '__main__':
    n, m = map(int, input().split())
    arr = list(map(int, input().split()))

    result = cutter(arr, m)
    print(result)